File: libraries/python/src/pandaprobe/tracing/span.py
"""SpanContext — manages the lifecycle of a single span within a trace."""

from __future__ import annotations

import logging
from datetime import datetime, timezone
from typing import TYPE_CHECKING, Any
from uuid import uuid4

from pandaprobe.schemas import SpanData, SpanKind, SpanStatusCode
from pandaprobe.tracing.context import get_span_stack
from pandaprobe.validation import validate_span_input, validate_span_output

if TYPE_CHECKING:
    from pandaprobe.tracing.context import TraceContext

logger = logging.getLogger("pandaprobe")


class SpanContext:
    """Context manager for an individual span.

    For LLM spans (kind="LLM"), input must be
    ``{"messages": [{"role": "...", "content": "..."}]}`` (full conversation
    history) and output must be ``{"messages": [{"role": "...", "content": "..."}]}``.
    Other span kinds accept arbitrary input/output.

    Automatically manages timing, parent-child relationships via the span
    stack (contextvar), and error capture.
    """

    def __init__(
        self,
        trace_ctx: TraceContext,
        name: str,
        *,
        kind: SpanKind = SpanKind.OTHER,
        model: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> None:
        self._trace_ctx = trace_ctx
        self._span_id = str(uuid4())
        self._name = name
        self._kind = kind
        self._model = model
        self._metadata = metadata or {}
        self._input: Any = None
        self._output: Any = None
        self._status: SpanStatusCode = SpanStatusCode.UNSET
        self._error: str | None = None
        self._started_at: datetime | None = None
        self._ended_at: datetime | None = None
        self._token_usage: dict[str, int] | None = None
        self._model_parameters: dict[str, Any] | None = None
        self._cost: dict[str, float] | None = None
        self._completion_start_time: datetime | None = None
        self._parent_span_id: str | None = None

    # ------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------

    @property
    def span_id(self) -> str:
        return self._span_id

    # ------------------------------------------------------------------
    # Context manager
    # ------------------------------------------------------------------

    def __enter__(self) -> SpanContext:
        self._started_at = datetime.now(timezone.utc)
        stack = get_span_stack()
        if stack:
            self._parent_span_id = stack[-1]
        stack.append(self._span_id)
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:  # noqa: ANN001
        self._ended_at = datetime.now(timezone.utc)
        if exc_type is not None:
            self._status = SpanStatusCode.ERROR
            self._error = str(exc_val)
        elif self._status == SpanStatusCode.UNSET:
            self._status = SpanStatusCode.OK

        self._finalize()

        stack = get_span_stack()
        if stack and stack[-1] == self._span_id:
            stack.pop()

        return None

    async def __aenter__(self) -> SpanContext:
        return self.__enter__()

    async def __aexit__(self, exc_type, exc_val, exc_tb):  # noqa: ANN001
        return self.__exit__(exc_type, exc_val, exc_tb)

    # ------------------------------------------------------------------
    # Setters
    # ------------------------------------------------------------------

    def set_input(self, input: Any) -> None:
        if self._kind == SpanKind.LLM:
            validate_span_input(input)
        self._input = input

    def set_output(self, output: Any) -> None:
        if self._kind == SpanKind.LLM:
            validate_span_output(output)
        self._output = output

    def set_token_usage(self, *, prompt_tokens: int = 0, completion_tokens: int = 0, **extra: int) -> None:
        self._token_usage = {"prompt_tokens": prompt_tokens, "completion_tokens": completion_tokens, **extra}

    def set_model(self, model: str) -> None:
        self._model = model

    def set_model_parameters(self, params: dict[str, Any]) -> None:
        self._model_parameters = params

    def set_cost(self, *, total: float, **extra: float) -> None:
        self._cost = {"total": total, **extra}

    def set_completion_start_time(self, ts: datetime) -> None:
        self._completion_start_time = ts

    def set_error(self, error: str) -> None:
        self._error = error
        self._status = SpanStatusCode.ERROR

    def set_metadata(self, metadata: dict[str, Any]) -> None:
        self._metadata.update(metadata)

    # ------------------------------------------------------------------
    # Finalization
    # ------------------------------------------------------------------

    def _finalize(self) -> None:
        try:
            span = SpanData(
                span_id=self._span_id,
                parent_span_id=self._parent_span_id,
                name=self._name,
                kind=self._kind,
                status=self._status,
                input=self._input,
                output=self._output,
                model=self._model,
                token_usage=self._token_usage,
                metadata=self._metadata,
                started_at=self._started_at,
                ended_at=self._ended_at,
                error=self._error,
                completion_start_time=self._completion_start_time,
                model_parameters=self._model_parameters,
                cost=self._cost,
            )
            self._trace_ctx.add_span_data(span)
        except Exception as exc:
            logger.error("PandaProbe failed to finalize span: %s", exc)

File: libraries/typescript/src/tracing/span.ts
/**
 * SpanContext — manages the lifecycle of a single span within a trace.
 *
 * For LLM spans (kind="LLM"), input/output must follow the messages schema.
 * Other span kinds accept arbitrary input/output. Parent-child relationships
 * are managed via the owning trace's own span stack (`traceCtx.spanStack`), so
 * spans always nest within their own trace regardless of the global context.
 */

import { logger } from "../logger.js";
import { SpanData, SpanKind, SpanStatusCode, newUuid } from "../schemas.js";
import { isThenable } from "../util.js";
import { warnIfInvalidMessages } from "../validation.js";
import type { TraceContext } from "./context.js";

export interface SpanContextOptions {
  kind?: SpanKind;
  model?: string | null;
  metadata?: Record<string, unknown>;
}

export class SpanContext {
  private readonly traceCtx: TraceContext;
  private readonly _spanId: string;
  private readonly name: string;
  private readonly kind: SpanKind;
  private model: string | null;
  private readonly metadata: Record<string, unknown>;
  private input: unknown = null;
  private output: unknown = null;
  private status: SpanStatusCode = SpanStatusCode.UNSET;
  private error: string | null = null;
  private startedAt: Date | null = null;
  private endedAt: Date | null = null;
  private tokenUsage: Record<string, number> | null = null;
  private modelParameters: Record<string, unknown> | null = null;
  private cost: Record<string, number> | null = null;
  private completionStartTime: Date | null = null;
  private parentSpanId: string | null = null;

  constructor(traceCtx: TraceContext, name: string, options: SpanContextOptions = {}) {
    this.traceCtx = traceCtx;
    this._spanId = newUuid();
    this.name = name;
    this.kind = options.kind ?? SpanKind.OTHER;
    this.model = options.model ?? null;
    this.metadata = options.metadata ?? {};
  }

  get spanId(): string {
    return this._spanId;
  }

  /** The output recorded on this span (used to mirror it onto standalone traces). */
  getOutput(): unknown {
    return this.output;
  }

  /** The trace this span belongs to (used to attach sibling/child tool spans). */
  getTraceContext(): TraceContext {
    return this.traceCtx;
  }

  // ------------------------------------------------------------------
  // Lifecycle
  // ------------------------------------------------------------------

  /** Enter the span: record start time, set parent, push onto the trace's stack. */
  start(): this {
    this.startedAt = new Date();
    const stack = this.traceCtx.spanStack;
    if (stack.length > 0) {
      this.parentSpanId = stack[stack.length - 1] ?? null;
    }
    stack.push(this._spanId);
    return this;
  }

  /** End the span: record end time/status, finalize, pop from the trace's stack. */
  end(error?: unknown): void {
    this.endedAt = new Date();
    if (error !== undefined && error !== null) {
      this.status = SpanStatusCode.ERROR;
      this.error = String(error instanceof Error ? error.message : error);
    } else if (this.status === SpanStatusCode.UNSET) {
      this.status = SpanStatusCode.OK;
    }
    this.finalize();
    const stack = this.traceCtx.spanStack;
    if (stack.length > 0 && stack[stack.length - 1] === this._spanId) {
      stack.pop();
    }
  }

  /** Run *fn* within this span (callback form). */
  async run<T>(fn: (span: SpanContext) => T | Promise<T>): Promise<T> {
    this.start();
    let errored: unknown;
    try {
      return await fn(this);
    } catch (exc) {
      errored = exc;
      throw exc;
    } finally {
      this.end(errored);
    }
  }

  /**
   * Wrap a sync-or-async invocation in this span (decorator form). Sets the
   * given *input*, captures the return value as output, and preserves the
   * original return type (sync stays sync, async stays async).
   */
  instrument<T>(invoke: () => T, input: unknown): T {
    this.start();
    this.setInput(input);
    let result: T;
    try {
      result = invoke();
    } catch (exc) {
      this.end(exc);
      throw exc;
    }
    if (isThenable(result)) {
      const wrapped = (result as unknown as Promise<unknown>).then(
        (value) => {
          this.setOutput(value);
          this.end();
          return value;
        },
        (exc) => {
          this.end(exc);
          throw exc;
        },
      );
      return wrapped as T;
    }
    this.setOutput(result);
    this.end();
    return result;
  }

  // ------------------------------------------------------------------
  // Setters
  // ------------------------------------------------------------------

  setInput(input: unknown): void {
    if (this.kind === SpanKind.LLM) {
      warnIfInvalidMessages(input, "LLM span input");
    }
    this.input = input;
  }

  setOutput(output: unknown): void {
    if (this.kind === SpanKind.LLM) {
      warnIfInvalidMessages(output, "LLM span output");
    }
    this.output = output;
  }

  setTokenUsage(usage: { promptTokens?: number; completionTokens?: number; [key: string]: number | undefined }): void {
    const { promptTokens = 0, completionTokens = 0, ...extra } = usage;
    const out: Record<string, number> = { prompt_tokens: promptTokens, completion_tokens: completionTokens };
    for (const [k, v] of Object.entries(extra)) {
      if (typeof v === "number") {
        out[k] = v;
      }
    }
    this.tokenUsage = out;
  }

  setModel(model: string): void {
    this.model = model;
  }

  setModelParameters(params: Record<string, unknown>): void {
    this.modelParameters = params;
  }

  setCost(cost: { total: number; [key: string]: number }): void {
    this.cost = { ...cost };
  }

  setCompletionStartTime(ts: Date): void {
    this.completionStartTime = ts;
  }

  setError(error: string): void {
    this.error = error;
    this.status = SpanStatusCode.ERROR;
  }

  setMetadata(metadata: Record<string, unknown>): void {
    Object.assign(this.metadata, metadata);
  }

  // ------------------------------------------------------------------
  // Finalization
  // ------------------------------------------------------------------

  private finalize(): void {
    try {
      const span = new SpanData({
        spanId: this._spanId,
        parentSpanId: this.parentSpanId,
        name: this.name,
        kind: this.kind,
        status: this.status,
        input: this.input,
        output: this.output,
        model: this.model,
        tokenUsage: this.tokenUsage,
        metadata: this.metadata,
        startedAt: this.startedAt ?? new Date(),
        endedAt: this.endedAt,
        error: this.error,
        completionStartTime: this.completionStartTime,
        modelParameters: this.modelParameters,
        cost: this.cost,
      });
      this.traceCtx.addSpanData(span);
    } catch (exc) {
      logger.error(`failed to finalize span: ${String(exc)}`);
    }
  }
}
